Serialize outgoing peer requests

jcramer opened the pull request:
We noticed on a testnet with 1 or 2 other available peers the node would make many (up to 8) connection requests to the same peer. This resulted in our node getting banned by the other peers because of the spamming.

Simply serializing the call for NewConnReq significantly reduces duplicate connections. Sometimes we still see a duplicate connection on startup, probably due to some race condition in the call cm.cfg.GetNewAddress.

This fixes the startup related issue https://github.com/gcash/bchd/issues/448, and may also fix https://github.com/gcash/bchd/pull/296 but not sure the entire context for this PR.

zquestz commented on connmgr/connmanager.go:
```diff
@@ -526,9 +526,11 @@ func (cm *ConnManager) Start() {
-	for i := atomic.LoadUint64(&cm.connReqCount); i < uint64(cm.cfg.TargetOutbound); i++ {
-		go cm.NewConnReq()
-	}
+	go func() {
```
Not sure what this is meant to accomplish. This is just wrapping `cm.NewConnReq()` in a go routine that executes immediately. I don't see anything that would prevent duplicate connections. I was expecting a mutex or some other lock to make them go serially.

jcramer commented:
I moved the go routine from inside the for loop to outside the for loop so
we’re not spamming the NewConnReq call.  We could alternatively put a lock
inside NeeConnReq if you all want that instead.

I did have the lock, but then I changed it.

On Wed, Mar 10, 2021 at 10:50 PM Josh Ellithorpe <notifications@github.com>
wrote:

> *@zquestz* commented on this pull request.
> ------------------------------
>
> In connmgr/connmanager.go
> <https://github.com/gcash/bchd/pull/449#discussion_r592047851>:
>
> > @@ -526,9 +526,11 @@ func (cm *ConnManager) Start() {
>  		}
>  	}
>
> -	for i := atomic.LoadUint64(&cm.connReqCount); i < uint64(cm.cfg.TargetOutbound); i++ {
> -		go cm.NewConnReq()
> -	}
> +	go func() {
>
> Not sure what this is meant to accomplish. This is just wrapping
> cm.NewConnReq() in a go routine that executes immediately. I don't see
> anything that would prevent duplicate connections. I was expecting a mutex
> or some other lock to make them go serially.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gcash/bchd/pull/449#pullrequestreview-609373555>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHRIIYQ4AJMUE2WMBAXZIDTDA4YLANCNFSM4Y7MR5CQ>
> .
>

jcramer commented:
I'll update it to be a lock so it's serial all the time, not only at start up as this does.

jcramer commented:
@zquestz this has been updated to use a lock instead.
